### ISSUE TITLE
Creating a StreamProvider from a FutureProvider

### ISSUE THREAD

**Issue by funwithflutter** [rrousselGit/river_pod#122]
The documentation has examples showing how to create a FutureProvider from a StreamProvider, but not the other way around. It's not entirely clear how providers can be built from a FutureProvider, unless you make those providers return AsyncValue. I've taken a look at all of the documentation, and example applications, but can't come up with a clean solution for this problem.

Please note though that I know there are other ways to go about this, but I'd like to know if something like this is even possible. The example I'm using is to create a "Box" in Hive for local storage.

```dart
// Please comment on whether I'm using autoDispose correctly, but regardles of 
// that, pretend that this is any kind of Future (with or without autoDispose).
final boxProvider = FutureProvider.autoDispose<Box>((ref) async {
  final box = await Hive.openBox('testBox');
  ref.onDispose(() {
    box.close();
  });
  return box;
});

// Hive allows you to return a stream of BoxEvents. How can I expose that stream, from the FutureProvider,
// that other providers can then watch for changes
final boxStream = StreamProvider.autoDispose<BoxEvent>((ref) {
  final box = ref.watch(boxProvider);
  return box.whenData((value) => value.watch()); // Which returns AsyncValue<Stream<BoxEvent>>
});

// And pretend `ListenableProvider` exists for demonstration purposes. The same applies
final boxListenable = ListenableProvider.autoDispose<Box>((ref) {
  final box = ref.watch(boxProvider);
  return box.whenData((value) => value.listenable());
});
```

**Another question:**
Is there a clean way to force a provider to execute, or to be non-lazy.

In the context of this example, Hive supports `Hive.box('testBox')`, which returns a singleton of an already open box. So it's possible to do something like this:

```dart
final boxStream = StreamProvider.autoDispose<BoxEvent>((ref) {
  // final box = ref.watch(boxProvider);
  final box =Hive.box('testBox');
  return box.watch();
});
```
Then you can proceed to use `boxStream` in your UI. However, in this instance the original `boxProvider` is no longer used and as a result that step where the box is opened won't happen. Unless you call this provider from the UI.

To sum up. How would you approach a situation like this. And is there a way to force a provider to do some form of 'initialisation`.

Really been enjoying Riverpod! This has been my only stumbling block, and maybe it's just my head that's the hurdle.

**Comment by rrousselGit** [comment]
Your boxProvider/boxStream are correct. There's not much to say about it

> Is there a clean way to force a provider to execute, or to be non-lazy.

Simply read the provider anywhere in your app. It could be in the main or MyApp. That will force it to execute

**Comment by funwithflutter** [comment]
Just to be clear, this boxProvider, does not have the correct return type:

```dart
final boxStream = StreamProvider.autoDispose<BoxEvent>((ref) {
  final box = ref.watch(boxProvider);
  return box.whenData((value) => value.watch()); // Which returns AsyncValue<Stream<BoxEvent>>
});
```

**Comment by rrousselGit** [comment]
Ah I missed it. You can do:

```dart
final boxStream = StreamProvider.autoDispose<BoxEvent>((ref) async* {
  final box = await ref.watch(boxProvider.future);
  yield* box.watch());
});
```

**Comment by funwithflutter** [comment]
Legend. Thanks for that.

**Comment by rrousselGit** [comment]
Can we close this then? :smile:
